[STORM-1273]  port backtype.storm.cluster to java

hustfxj opened the pull request:
1. port cluster.clj to ClusterUtils/StormClusterState/StateStorage/ZKStateStorage/ZKStateStorageFactory
2. Update all the callings to cluster
3. Mock the java static functions for test purposes.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/cluster/ClusterUtils.java:
```diff
@@ -0,0 +1,249 @@
+    }
+
+    public StormClusterState mkStormClusterStateImpl(Object StateStorage, List<ACL> acls, ClusterStateContext context) throws Exception {
+        if (StateStorage instanceof StateStorage) {
```
can you rename StateStorage variable to stateStorage (camel case)

hustfxj replied:
ok

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/cluster/ClusterUtils.java:
```diff
@@ -0,0 +1,249 @@
+        Map<ExecutorInfo, ExecutorStats> executorStatsMap = workerHeartbeat.get_executor_stats();
+        for (ExecutorInfo executor : executors) {
+            if (executorStatsMap.containsKey(executor)) {
+                executorWhb.put(executor, workerHeartbeat);
```
in equivalent clojure code, (executor-stats, uptime) is being put into the map instead of same worker heart beat object. is it intentional?

hustfxj replied:
yes, It is intentional to return  ClusterWorkerHeartbeat which can be converted clojure code by clojurify-zk-worker-hb. So it can be called by  clojure.

abhishekagarwal87 replied:
result map value here doesn't have executor-stats and executor uptime but the whole clusterWorkerHeartBeat. It's not the object type and structure I refer to but the data in object.

hustfxj replied:
thank you, I  fixed it

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/cluster/StormClusterStateImpl.java:
```diff
@@ -0,0 +1,664 @@
+            }
+            Collections.sort(errorInfos, new Comparator<ErrorInfo>() {
+                public int compare(ErrorInfo arg0, ErrorInfo arg1) {
+                    return -Integer.compare(arg0.get_error_time_secs(), arg1.get_error_time_secs());
```
Instead of minus, you can reorder the arguments.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/cluster/ClusterUtils.java:
```diff
@@ -0,0 +1,249 @@
+
+    }
+
+    public StateStorage mkDistributedClusterStateImpl(APersistentMap config, APersistentMap auth_conf, List<ACL> acls, ClusterStateContext context)
```
this can be renamed to mkStateStorageImpl.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/cluster/StateStorageFactory.java:
```diff
@@ -21,8 +21,8 @@
+public interface StateStorageFactory {
     
-    ClusterState mkState(APersistentMap config, APersistentMap auth_conf, List<ACL> acls, ClusterStateContext context);
+    StateStorage mkState(APersistentMap config, APersistentMap auth_conf, List<ACL> acls, ClusterStateContext context);
```
function should be renamed to mkStore

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/cluster/ZKStateStorage.java:
```diff
@@ -0,0 +1,271 @@
+
+    @Override
+    public void delete_node_blobstore(String path, String nimbusHostPortInfo) {
+
```
not implemented.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/cluster/StateStorage.java:
```diff
@@ -38,16 +41,16 @@
  * Never use the same paths with the *_hb* methods as you do with the others.
  */
-public interface ClusterState {
+public interface StateStorage {
```
StateStorage refers to classes which are related to zookeeper. If you want complete abstraction over different state store implementation, you should remove zookeeper specific dependencies in the interface.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/cluster/StormClusterState.java:
```diff
@@ -0,0 +1,129 @@
+import java.util.List;
+import java.util.Map;
+
+public interface StormClusterState {
```
all interfaces name should be prefix with I e.g. IStormClusterState, IStateStorage

HeartSaVioR replied:
Naming rule for interface seems not that consistent. We may want to address naming convention, but it needs to gather consensus.

abhishekagarwal87 replied:
I would definitely vote for it. It becomes easier to browse and understand the code.

hustfxj replied:
yes， you are right. I will fix it

abhishekagarwal87 commented on storm-core/src/clj/org/apache/storm/daemon/nimbus.clj:
```diff
@@ -587,7 +586,12 @@
+        executor-beats (let [executor-stats-java-map (.executorBeats storm-cluster-state storm-id (.get_executor_node_port (thriftify-assignment existing-assignment)))
+                             executor-stats-clojurify (clojurify-structure executor-stats-java-map)]
+                         (->> (dofor [[^ExecutorInfo executor-info ^ClusterWorkerHeartbeat cluster-worker-heartbeat] executor-stats-clojurify]
+                             {[(.get_task_start executor-info) (.get_task_end executor-info)] (clojurify-zk-worker-hb cluster-worker-heartbeat)})
```
with reference #diff-3dc2ee75946dae2afc9de7d4e5487219R191
in earlier version,
the structure returned was {["task_start", "task_end"] : {time-secs, uptime, executor-stat}}
now the structure returned seems to be {["task_start", "task_end"] : {storm-id, executor-stats, time-secs, uptime-secs}}
It seems for each executor, clojurified cluster worker heart beat is being stored. can you confirm this is not the case?

hustfxj replied:
ok thank you for your comments. I get it.

hustfxj commented:
@revans2 @abhishekagarwal87  can you help review these commits ?

zhuoliu commented:
We would also like @knusbaum to review this one.

revans2 commented on storm-core/src/clj/org/apache/storm/daemon/supervisor.clj:
```diff
@@ -1183,7 +1184,8 @@
       (finally
         (.shutdown blob-store)))
-    (FileUtils/moveDirectory (File. tmproot) (File. stormroot))
+    (try (FileUtils/moveDirectory (File. tmproot) (File. stormroot)) (catch Exception e))
```
Why are we ignoring this exception now?

hustfxj replied:
Sorry, I forget remove the “try”.   It can't pass testing at "FileUtils/moveDirectory" because the windows operating system. I will ignore this exception per test.

revans2 commented on storm-core/src/clj/org/apache/storm/testing.clj:
```diff
@@ -447,8 +447,8 @@
-        assignment (.assignment-info state storm-id nil)
-        taskbeats (.taskbeats state storm-id (:task->node+port assignment))
+        assignment (clojurify-assignment (.assignmentInfo state storm-id nil))
+        taskbeats (.taskbeats state storm-id (:task->node+port assignment))  ;hava question?
```
What is your question here?

hustfxj replied:
forget remove the comments. Thanks for your reminder.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/ClusterUtils.java:
```diff
@@ -0,0 +1,282 @@
+
+import clojure.lang.APersistentMap;
+import clojure.lang.PersistentArrayMap;
+import clojure.lang.RT;
```
We should not depend on any clojure internals.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/ClusterUtils.java:
```diff
@@ -0,0 +1,282 @@
+     * @param workerHeartbeat
+     * @return
+     */
+    public static Map<ExecutorInfo, APersistentMap> convertExecutorBeats(List<ExecutorInfo> executors, ClusterWorkerHeartbeat workerHeartbeat) {
```
These should just be maps, and if we need to clojurify them outside of this we can.

revans2 replied:
Actually the return value should probably be something like Map<ExecutorInfo, ExecutorBeat> where ExecutorBeat is something like.

```
public class ExecutorBeat {
    public final long timeSecs;
    public final long uptime;
    public final ExecutorStats stats;
}
```

Then you can have something translate it to the clojure map if needed

hustfxj replied:
should ExecutorBeat  be generated by thrift ?

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/ClusterUtils.java:
```diff
@@ -0,0 +1,282 @@
+
+    }
+
+    public IStateStorage mkStateStorageImpl(APersistentMap config, APersistentMap auth_conf, List<ACL> acls, ClusterStateContext context) throws Exception {
```
APersistentMap should just be Map here

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/ClusterUtils.java:
```diff
@@ -0,0 +1,282 @@
+        return stateStorage;
+    }
+
+    public static IStateStorage mkStateStorage(APersistentMap config, APersistentMap auth_conf, List<ACL> acls, ClusterStateContext context) throws Exception {
```
APersistentMap should just be Map here

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/ClusterUtils.java:
```diff
@@ -0,0 +1,282 @@
+    }
+
+    // TO be remove
+    public static <K, V> HashMap<V, List<K>> reverseMap(Map<K, V> map) {
```
Please remove this there is a version in Utils already.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/ClusterUtils.java:
```diff
@@ -0,0 +1,282 @@
+        return rtn;
+    }
+
+    public static String StringifyError(Throwable error) {
```
Can we make it be a lower case 's' so it follows the java naming convention.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/IStateStorage.java:
```diff
@@ -18,12 +18,14 @@
 package org.apache.storm.cluster;
 
 import clojure.lang.APersistentMap;
```
No clojure internals here either please.

revans2 replied:
Because you didn't add this, we can do it in a follow in JIRA if you want.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/StormClusterStateImpl.java:
```diff
@@ -0,0 +1,687 @@
+    protected void issueCallback(AtomicReference<IFn> cb) {
+        IFn callback = cb.getAndSet(null);
+        if (callback != null)
+            callback.invoke();
```
Anything that calls invoke with no parameters should be a Runnable not an IFn.

revans2 commented:
This is a really big pull request and things are changing all of the time with others merging things in.  Perhaps we should break this up into smaller pieces.  Like just doing each state factory one at a time, and then some of the other pieces?  IF you want to keep this all together that is OK, but it is going to take longer to really review it and get it in.

hustfxj commented:
@revans2  thanks for your review, I have addressed your comments and removed  any clojure internals.
I want to keep this all together. I will keep merging with the master all time. Of course, I really hope  every one can review it and help me get it in.

revans2 commented:
@hustfxj OK I will try to get through the entire thing today.  So we can get it in soon.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/ClusterUtils.java:
```diff
@@ -0,0 +1,261 @@
+    public static List<ACL> mkTopoOnlyAcls(Map topoConf) throws NoSuchAlgorithmException {
+        List<ACL> aclList = null;
+        String payload = (String) topoConf.get(Config.STORM_ZOOKEEPER_TOPOLOGY_AUTH_PAYLOAD);
+        if (Utils.isZkAuthenticationConfiguredStormServer(topoConf)) {
```
This needs to be isZkAuthenticationConfiguredTopology not isZkAuthenticationConfiguredStormServer.
Original code is 
`(when (Utils/isZkAuthenticationConfiguredTopology topo-conf)`

hustfxj replied:
@revans2  Thank you. I have addressed your comments.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/ClusterUtils.java:
```diff
@@ -0,0 +1,261 @@
+        return executorWhb;
+    }
+
+    public IStormClusterState mkStormClusterStateImpl(Object stateStorage, List<ACL> acls, ClusterStateContext context) throws Exception {
```
Can we split this and mkStormClusterState into two methods?  The signature of each are different enough that having them be separate is preferable.

hustfxj replied:
I agree with you. But it should be update into two methods after finishing some tests. Because it may happen some problems at tests  when it has two methods.For example, (.mkStormClusterStateImpl (Mockito/verify cluster-utils (Mockito/times 1)) (Mockito/any) (Mockito/eq expected-acls) (Mockito/any))

revans2 replied:
OK we can leave it for now.  But please file a follow on JIRA so we can clean it up later.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/ClusterUtils.java:
```diff
@@ -0,0 +1,261 @@
+            }
+        }
+        return errorString;
+    }
```
This can be made a lot simpler with the java 7 AutoClosable semantics.

```
    public static String stringifyError(Throwable error) {
        try (
            StringWriter result = new StringWriter();;
            PrintWriter printWriter = new PrintWriter(result);
        ){
            error.printStackTrace(printWriter);
            return result.toString();
        }
    }
```

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/PaceMakerStateStorage.java:
```diff
@@ -0,0 +1,212 @@
+                if (retry <= 0) {
+                    throw Utils.wrapInRuntime(e);
+                }
+                LOG.error("{} Failed to set_worker_hb. Will make {} more attempts.", e.getMessage(), retry--);
```
Lets move decrementing retry outside of the log message.  It is too easy to delete/comment out a log message and if there are side effects it becomes a problem.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/PaceMakerStateStorage.java:
```diff
@@ -0,0 +1,212 @@
+                if (retry <= 0) {
+                    throw Utils.wrapInRuntime(e);
+                }
+                LOG.error("{} Failed to get_worker_hb. Will make {} more attempts.", e.getMessage(), retry--);
```
Same here, no side effects from a log message please.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/PaceMakerStateStorage.java:
```diff
@@ -0,0 +1,212 @@
+                if (retry <= 0) {
+                    throw Utils.wrapInRuntime(e);
+                }
+                LOG.error("{} Failed to get_worker_hb_children. Will make {} more attempts.", e.getMessage(), retry--);
```
Same here, no side effects from a log message please.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/PaceMakerStateStorage.java:
```diff
@@ -0,0 +1,212 @@
+                if (retry <= 0) {
+                    throw Utils.wrapInRuntime(e);
+                }
+                LOG.error("{} Failed to delete_worker_hb. Will make {} more attempts.", e.getMessage(), retry--);
```
Same here, no side effects from a log message please.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/StormClusterStateImpl.java:
```diff
@@ -0,0 +1,687 @@
+                    } else if (root.equals(ClusterUtils.LOGCONFIG_ROOT) && size > 1) {
+                        issueMapCallback(logConfigCallback, toks.get(1));
+                    } else if (root.equals(ClusterUtils.BACKPRESSURE_ROOT) && size > 1) {
+                        issueMapCallback(logConfigCallback, toks.get(1));
```
This should be backpressureCallback.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/StormClusterStateImpl.java:
```diff
@@ -0,0 +1,687 @@
+    protected void issueMapCallback(ConcurrentHashMap<String, Runnable> callbackConcurrentHashMap, String key) {
+        Runnable callback = callbackConcurrentHashMap.remove(key);
+        if (callback != null)
+            callback.run();
```
In the previous code issue-map-callback passed in a parameter.  The key.  We are not doing that any more, but it looked like everywhere that was ignored.  This is fine, but for clarity we should just remove the `& ignored` everywhere in the callback functions.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/IStormClusterState.java:
```diff
@@ -0,0 +1,122 @@
+
+    public List<ProfileRequest> getWorkerProfileRequests(String stormId, NodeInfo nodeInfo, boolean isThrift);
+
+    public List<ProfileRequest> getTopologyProfileRequests(String stormId, boolean isThrift);
```
The isThrift is ignored by all of the implementations.  We should just remove it if we are not going to support it.

revans2 replied:
Actually it looks like it is true in some places and false in others.  We need to either implement it or make it so it is always true and remove it.

revans2 replied:
OK I see that it is not needed in either of the places it is called, so yes, lets remove it.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/StormClusterStateImpl.java:
```diff
@@ -0,0 +1,687 @@
+
+    @Override
+    public LogConfig topologyLogConfig(String stormId, Runnable cb) {
+        String path = ClusterUtils.logConfigPath(stormId);
```
cb is never stored to be called back.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/StormClusterStateImpl.java:
```diff
@@ -0,0 +1,687 @@
+                    return Integer.compare(arg1.get_error_time_secs(), arg0.get_error_time_secs());
+                }
+            });
+        } catch (Exception e) {
```
The original code was not wrapping the Exception in a RuntimeException

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/ZKStateStorage.java:
```diff
@@ -0,0 +1,270 @@
+                    if (!(state.equals(Watcher.Event.KeeperState.SyncConnected))) {
+                        LOG.warn("Received event {} : {}: {} with disconnected Zookeeper.", state, type, path);
+                    } else {
+                        LOG.info("Received event {} : {} : {}", state, type, path);
```
This info message was not logged before. Could we at least make it debug.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/ZKStateStorage.java:
```diff
@@ -0,0 +1,270 @@
+            }
+        });
+        if (isNimbus) {
+            zkReader = mkZk(new WatcherCallBack() {
```
It might be nice to reuse the same watcher callback instance, just so we don't have two identical anonymous classes.

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/ZKStateStorage.java:
```diff
@@ -0,0 +1,270 @@
+                if (Utils.exceptionCauseIsInstanceOf(KeeperException.NoNodeException.class, e)) {
+                    Zookeeper.createNode(zkWriter, path, data, CreateMode.EPHEMERAL, acls);
+                } else {
+                    throw Utils.wrapInRuntime(e);
```
The original code did not wrap this in a Runtime Exception

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/ZKStateStorage.java:
```diff
@@ -0,0 +1,270 @@
+
+    @Override
+    public boolean node_exists(String path, boolean watch) {
+        return Zookeeper.existsNode(zkWriter, path, watch);
```
This should be zkReader, not zkWriter

revans2 commented on storm-core/src/jvm/org/apache/storm/cluster/ZKStateStorage.java:
```diff
@@ -0,0 +1,270 @@
+    }
+
+    // To be remove when finished port Util.clj
+    public static String parentPath(String path) {
```
This can be removed now.

hustfxj replied:
Don't transalte parentPath at Utils.clj. So I will move the func to Utils.clj

revans2 commented:
OK I made it all of the way through.  Things look fairly good, but there are a few places that need to be updated/fixed.

hustfxj commented:
@revans2  Thank you again. I have addressed your comments.

revans2 commented:
+1 looks god to me.  Did some manual testing and things look good.
